File: app/src/main/java/org/nevack/unitconverter/model/converter/CurrencyConverter.java
package org.nevack.unitconverter.model.converter;

import android.content.Context;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import org.nevack.unitconverter.NBRBService;
import org.nevack.unitconverter.R;
import org.nevack.unitconverter.model.Rate;
import org.nevack.unitconverter.model.Unit;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.reflect.Type;
import java.util.Calendar;
import java.util.Collections;
import java.util.List;

import retrofit2.Call;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class CurrencyConverter extends Converter {
    private static final String FILE = "rates.json";

    private static final Unit BYN = new Unit("Белорусский рубль", 1d, "BYN");
    private final Gson gson;
    private final File file;

    public CurrencyConverter(Context context) {

        gson = new Gson();

        file = new File(context.getFilesDir(), FILE);
        if (file.exists()) {
            Calendar calendar = Calendar.getInstance();
            calendar.setTimeInMillis(file.lastModified());
            if (Calendar.getInstance().get(Calendar.DAY_OF_YEAR) == calendar.get(Calendar.DAY_OF_YEAR)) {
                loadUnitsFromFile();
            } else {
                loadUnitsFromWeb();
            }
        } else {
            loadUnitsFromWeb();
        }

        units.add(BYN);

        Collections.sort(units, (lhs, rhs) -> lhs.getName().compareToIgnoreCase(rhs.getName()));
    }

    private void loadUnitsFromFile() {
        try {
            Type type = new TypeToken<List<Rate>>(){}.getType();
            List<Rate> rates = gson.fromJson(new FileReader(file), type);
            for (Rate rate : rates) {
                units.add(rate.toUnit());
            }
        } catch (FileNotFoundException ex) {
            ex.printStackTrace();
        }
    }

    private void loadUnitsFromWeb() {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://www.nbrb.by/API/ExRates/")
                .addConverterFactory(GsonConverterFactory.create(gson)).build();

        NBRBService service = retrofit.create(NBRBService.class);
        Call<List<Rate>> call = service.getAllRatesForToday();

        try {
            Response<List<Rate>> response = call.execute();
            if (response.isSuccessful()) {
                List<Rate> rates = response.body();
                if (rates != null) {
                    for (Rate rate : rates) {
                        units.add(rate.toUnit());
                    }
                }

                Type type = new TypeToken<List<Rate>>() {}.getType();
                String json = gson.toJson(rates, type);
                FileWriter fw = new FileWriter(file);
                BufferedWriter out = new BufferedWriter(fw);
                out.write(json);
                out.close();
            }
        } catch (IOException e) {
            e.printStackTrace();
            loadUnitsFromFile();
        }
    }

    @Override
    public int getName() {
        return R.string.currency;
    }
}

File: app/src/main/java/org/nevack/unitconverter/model/converter/CurrencyConverter.kt
package org.nevack.unitconverter.model.converter

import android.content.Context
import com.squareup.moshi.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okio.buffer
import okio.sink
import okio.source
import org.nevack.unitconverter.NBRBService
import org.nevack.unitconverter.R
import org.nevack.unitconverter.model.Rate
import org.nevack.unitconverter.model.ConversionUnit
import retrofit2.Call
import java.io.*
import java.util.*

class CurrencyConverter(context: Context) : Converter(R.string.currency) {
    internal var service: NBRBService? = null
    internal var moshi: Moshi? = null
    private val adapter: JsonAdapter<List<Rate>> by lazy(LazyThreadSafetyMode.NONE) {
        requireNotNull(moshi).adapter(TYPE)
    }

    private val file: File = File(context.filesDir, FILE)

    override suspend fun load() {
        registerUnit(BYN)
        if (file.exists()) {
            val calendar = Calendar.getInstance()
            calendar.timeInMillis = file.lastModified()
            if (Calendar.getInstance()[Calendar.DAY_OF_YEAR] == calendar[Calendar.DAY_OF_YEAR]) {
                loadUnitsFromFile()
            } else {
                loadUnitsFromWeb()
            }
        } else {
            loadUnitsFromWeb()
        }
        sortUnitsWith { a, b -> a.name.compareTo(b.name, ignoreCase = true) }
    }

    private suspend fun loadUnitsFromFile() {
        val rates = withContext(Dispatchers.IO) {
            adapter.fromJson(file.source().buffer())
        }
        rates?.forEach { registerUnit(it.toUnit()) }
    }

    private suspend fun loadUnitsFromWeb() {
        val service = requireNotNull(service) { "Service is not set!" }
        val rates = service.allRatesForToday()
        rates.forEach { registerUnit(it.toUnit()) }
        withContext(Dispatchers.IO) {
            adapter.toJson(file.sink().buffer(), rates)
        }
    }

    companion object {
        private val TYPE = Types.newParameterizedType(List::class.java, Rate::class.java)
        private const val FILE = "rates.json"
        private val BYN = ConversionUnit("Белорусский рубль", 1.0, "BYN")
    }
}